<html>
<!-- This HTML was generated by the make_actions_html routine please modify --->
<!-- the source/utils/make_actions_html/actions_source_info file            --->
<head>
  <title>Avida : List of Actions</title>
	<link rel="stylesheet" type="text/css" href="main.css" />
</head>
<body>

<div class="revision">
Revised 2006-09-03 DMB
<br>Generated Fri Feb 29 12:15:08 2008 by make_actions_html</div>

<p>
<a href="index.html">Return to the Index</a> &nbsp;|&nbsp;
<a href="events.html">The Events File</a> &nbsp;|&nbsp;
<a href="analyze.html">The Analyze File</a>
</p>
<hr />

<div align="center">
<h1>List of Actions</h1>
</div>

<a name="ActionCategories"><h2>Action Categories</h2></a>
<p>
There is a large library of actions available for scheduling as events.
Additionally, all of these actions can be used within analyze scripts.
Below you will find a listing of the high level groupings of these
actions, along with detailed sections for each them.
</p>


<dl>
<dt><a href="#PrintActions">Print</a></dt>
<dd>Print actions are the primary way of saving data from an Avida experiments.</dd>
<dt><a href="#PopulationActions">Population</a></dt>
<dd>Population actions modify the state of the population, and will actually change the course of the run.</dd>
<dt><a href="#EnvironmentActions">Environment</a></dt>
<dd>Actions that allow user to change properties of the environment, including resources.</dd>
<dt><a href="#SaveLoadActions">Save and Load</a></dt>
<dd>Actions that allow for saving and loading large data sets, such as full populations.</dd>
<dt><a href="#LandscapeActions">Landscape Analysis</a></dt>
<dd>Actions that use data from the current state of Avida, process it and then output the results.</dd>
<dt><a href="#DriverActions">Driver</a></dt>
<dd>Actions that allow user to control program execution, including experiment termination.</dd>
</dl>

<p>
For a brief overview of writing a new action, please see <a href="#CreateAction">Creating an Action</a> below.
</p>

<p>&nbsp;</p>
<h2>Alphabetical Listing of Available Actions</h2>

<table>
  <tr>
    <td valign="top">
      <a href="#AnalyzeLandscape">AnalyzeLandscape</a><br>
      <a href="#AnalyzePopulation">AnalyzePopulation</a><br>
      <a href="#apocalypse">apocalypse</a><br>
      <a href="#ChangeEnvironment">ChangeEnvironment</a><br>
      <a href="#compete_demes">compete_demes</a><br>
      <a href="#CompeteDemes">CompeteDemes</a><br>
      <a href="#compete_organisms">compete_organisms</a><br>
      <a href="#CompeteOrganisms">CompeteOrganisms</a><br>
      <a href="#connect_cells">connect_cells</a><br>
      <a href="#ConnectCells">ConnectCells</a><br>
      <a href="#CopyDeme">CopyDeme</a><br>
      <a href="#copy_deme">copy_deme</a><br>
      <a href="#DelayedDemeEvent">DelayedDemeEvent</a><br>
      <a href="#DeletionLandscape">DeletionLandscape</a><br>
      <a href="#detail_parasite_pop">detail_parasite_pop</a><br>
      <a href="#detail_pop">detail_pop</a><br>
      <a href="#detail_sex_pop">detail_sex_pop</a><br>
      <a href="#DisconnectCells">DisconnectCells</a><br>
      <a href="#disconnect_cells">disconnect_cells</a><br>
      <a href="#DivideDemes">DivideDemes</a><br>
      <a href="#DumpCellDataGrid">DumpCellDataGrid</a><br>
      <a href="#dump_donor_grid">dump_donor_grid</a><br>
      <a href="#DumpDonorGrid">DumpDonorGrid</a><br>
      <a href="#DumpEnergyGrid">DumpEnergyGrid</a><br>
      <a href="#DumpExecutionRatioGrid">DumpExecutionRatioGrid</a><br>
      <a href="#dump_fitness_grid">dump_fitness_grid</a><br>
      <a href="#DumpFitnessGrid">DumpFitnessGrid</a><br>
      <a href="#dump_genotype_grid">dump_genotype_grid</a><br>
      <a href="#DumpGenotypeIDGrid">DumpGenotypeIDGrid</a><br>
      <a href="#dump_historic_pop">dump_historic_pop</a><br>
      <a href="#dump_historic_sex_pop">dump_historic_sex_pop</a><br>
      <a href="#DumpLandscape">DumpLandscape</a><br>
      <a href="#dump_lineage_grid">dump_lineage_grid</a><br>
      <a href="#DumpLineageGrid">DumpLineageGrid</a><br>
      <a href="#dump_memory">dump_memory</a><br>
      <a href="#DumpMemory">DumpMemory</a><br>
      <a href="#dump_phenotype_grid">dump_phenotype_grid</a><br>
      <a href="#DumpPhenotypeIDGrid">DumpPhenotypeIDGrid</a><br>
      <a href="#dump_pop">dump_pop</a><br>
      <a href="#DumpPopulation">DumpPopulation</a><br>
      <a href="#dump_receiver_grid">dump_receiver_grid</a><br>
      <a href="#DumpReceiverGrid">DumpReceiverGrid</a><br>
      <a href="#DumpSleepGrid">DumpSleepGrid</a><br>
      <a href="#dump_task_grid">dump_task_grid</a><br>
      <a href="#DumpTaskGrid">DumpTaskGrid</a><br>
      <a href="#Echo">Echo</a><br>
      <a href="#echo">echo</a><br>
      <a href="#exit">exit</a><br>
      <a href="#Exit">Exit</a><br>
      <a href="#ExitAveGeneration">ExitAveGeneration</a><br>
      <a href="#ExitAveLineageLabelGreater">ExitAveLineageLabelGreater</a><br>
      <a href="#ExitAveLineageLabelLess">ExitAveLineageLabelLess</a><br>
      <a href="#ExitElapsedTime">ExitElapsedTime</a><br>
      <a href="#exit_if_ave_lineage_label_larger">exit_if_ave_lineage_label_larger</a><br>
      <a href="#exit_if_ave_lineage_label_smaller">exit_if_ave_lineage_label_smaller</a><br>
      <a href="#FullLandscape">FullLandscape</a><br>
      <a href="#genetic_distance_pop_dump">genetic_distance_pop_dump</a><br>
      <a href="#HillClimb">HillClimb</a><br>
      <a href="#inject">inject</a><br>
      <a href="#Inject">Inject</a><br>
      <a href="#inject_all">inject_all</a><br>
      <a href="#InjectAll">InjectAll</a><br>
      <a href="#inject_all_random_repro">inject_all_random_repro</a><br>
      <a href="#InjectDemes">InjectDemes</a><br>
      <a href="#InjectParasite">InjectParasite</a><br>
      <a href="#InjectParasitePair">InjectParasitePair</a><br>
      <a href="#inject_random">inject_random</a><br>
      <a href="#InjectRandom">InjectRandom</a><br>
      <a href="#InjectRange">InjectRange</a><br>
      <a href="#inject_range">inject_range</a><br>
      <a href="#InjectResource">InjectResource</a><br>
      <a href="#inject_resource">inject_resource</a><br>
      <a href="#inject_scaled_resource">inject_scaled_resource</a><br>
      <a href="#InjectScaledResource">InjectScaledResource</a><br>
      <a href="#inject_sequence">inject_sequence</a><br>
      <a href="#InjectSequence">InjectSequence</a><br>
      <a href="#InjectSequenceWDivMutRate">InjectSequenceWDivMutRate</a><br>
      <a href="#InsertionLandscape">InsertionLandscape</a><br>
      <a href="#join_grid_col">join_grid_col</a><br>
      <a href="#join_grid_row">join_grid_row</a><br>
      <a href="#JoinGridCol">JoinGridCol</a><br>
      <a href="#JoinGridRow">JoinGridRow</a><br>
      <a href="#KillProb">KillProb</a><br>
      <a href="#KillRate">KillRate</a><br>
    <td valign="top">
      <a href="#KillRectangle">KillRectangle</a><br>
      <a href="#kill_rectangle">kill_rectangle</a><br>
      <a href="#LoadClone">LoadClone</a><br>
      <a href="#load_clone">load_clone</a><br>
      <a href="#load_dump_file">load_dump_file</a><br>
      <a href="#LoadPopulation">LoadPopulation</a><br>
      <a href="#ModMutProb">ModMutProb</a><br>
      <a href="#MutationalNeighborhood">MutationalNeighborhood</a><br>
      <a href="#NewTrial">NewTrial</a><br>
      <a href="#new_trial">new_trial</a><br>
      <a href="#outflow_scaled_resource">outflow_scaled_resource</a><br>
      <a href="#OutflowScaledResource">OutflowScaledResource</a><br>
      <a href="#PairTestLandscape">PairTestLandscape</a><br>
      <a href="#PrecalcLandscape">PrecalcLandscape</a><br>
      <a href="#PredictNuLandscape">PredictNuLandscape</a><br>
      <a href="#PredictWLandscape">PredictWLandscape</a><br>
      <a href="#print_average_data">print_average_data</a><br>
      <a href="#PrintAverageData">PrintAverageData</a><br>
      <a href="#PrintCCladeCounts">PrintCCladeCounts</a><br>
      <a href="#PrintCCladeFitnessHistogram">PrintCCladeFitnessHistogram</a><br>
      <a href="#PrintCCladeRelativeFitnessHistogram">PrintCCladeRelativeFitnessHistogram</a><br>
      <a href="#PrintCellVisitsData">PrintCellVisitsData</a><br>
      <a href="#PrintCompetitionData">PrintCompetitionData</a><br>
      <a href="#print_count_data">print_count_data</a><br>
      <a href="#PrintCountData">PrintCountData</a><br>
      <a href="#PrintCurrentReactionData">PrintCurrentReactionData</a><br>
      <a href="#print_data">print_data</a><br>
      <a href="#PrintData">PrintData</a><br>
      <a href="#PrintDebug">PrintDebug</a><br>
      <a href="#print_deme_stats">print_deme_stats</a><br>
      <a href="#PrintDemeAllStats">PrintDemeAllStats</a><br>
      <a href="#PrintDemeAverageData">PrintDemeAverageData</a><br>
      <a href="#PrintDemeDonorStats">PrintDemeDonorStats</a><br>
      <a href="#PrintDemeReplicationData">PrintDemeReplicationData</a><br>
      <a href="#PrintDemeResourceStats">PrintDemeResourceStats</a><br>
      <a href="#PrintDemeSpacialEnergyStats">PrintDemeSpacialEnergyStats</a><br>
      <a href="#PrintDemeSpacialSleepStats">PrintDemeSpacialSleepStats</a><br>
      <a href="#PrintDemeStats">PrintDemeStats</a><br>
      <a href="#print_depth_histogram">print_depth_histogram</a><br>
      <a href="#PrintDepthHistogram">PrintDepthHistogram</a><br>
      <a href="#print_detailed_fitness_data">print_detailed_fitness_data</a><br>
      <a href="#PrintDetailedFitnessData">PrintDetailedFitnessData</a><br>
      <a href="#print_divide_mut_data">print_divide_mut_data</a><br>
      <a href="#PrintDivideMutData">PrintDivideMutData</a><br>
      <a href="#print_dom">print_dom</a><br>
      <a href="#print_dom_parasite">print_dom_parasite</a><br>
      <a href="#print_dominant_data">print_dominant_data</a><br>
      <a href="#PrintDominantData">PrintDominantData</a><br>
      <a href="#PrintDominantGenotype">PrintDominantGenotype</a><br>
      <a href="#PrintDominantParasiteGenotype">PrintDominantParasiteGenotype</a><br>
      <a href="#print_donation_stats">print_donation_stats</a><br>
      <a href="#print_error_data">print_error_data</a><br>
      <a href="#PrintErrorData">PrintErrorData</a><br>
      <a href="#print_genetic_distance_data">print_genetic_distance_data</a><br>
      <a href="#PrintGeneticDistanceData">PrintGeneticDistanceData</a><br>
      <a href="#PrintGenomicSiteEntropy">PrintGenomicSiteEntropy</a><br>
      <a href="#print_genotype_abundance_histogram">print_genotype_abundance_histogram</a><br>
      <a href="#print_genotype_map">print_genotype_map</a><br>
      <a href="#PrintGenotypeAbundanceHistogram">PrintGenotypeAbundanceHistogram</a><br>
      <a href="#PrintGenotypeMap">PrintGenotypeMap</a><br>
      <a href="#print_genotypes">print_genotypes</a><br>
      <a href="#PrintGenotypes">PrintGenotypes</a><br>
      <a href="#PrintGermlineData">PrintGermlineData</a><br>
      <a href="#print_instruction_abundance_histogram">print_instruction_abundance_histogram</a><br>
      <a href="#print_instruction_data">print_instruction_data</a><br>
      <a href="#PrintInstructionAbundanceHistogram">PrintInstructionAbundanceHistogram</a><br>
      <a href="#PrintInstructionData">PrintInstructionData</a><br>
      <a href="#print_lineage_counts">print_lineage_counts</a><br>
      <a href="#print_lineage_totals">print_lineage_totals</a><br>
      <a href="#PrintLineageCounts">PrintLineageCounts</a><br>
      <a href="#PrintLineageTotals">PrintLineageTotals</a><br>
      <a href="#PrintLogFitnessHistogram">PrintLogFitnessHistogram</a><br>
      <a href="#print_market_data">print_market_data</a><br>
      <a href="#PrintMarketData">PrintMarketData</a><br>
      <a href="#print_mutation_rate_data">print_mutation_rate_data</a><br>
      <a href="#PrintMutationRateData">PrintMutationRateData</a><br>
      <a href="#print_number_phenotypes">print_number_phenotypes</a><br>
      <a href="#PrintParasiteData">PrintParasiteData</a><br>
      <a href="#print_phenotype_status">print_phenotype_status</a><br>
      <a href="#PrintPhenotypeData">PrintPhenotypeData</a><br>
      <a href="#PrintPhenotypeStatus">PrintPhenotypeStatus</a><br>
      <a href="#PrintPhenotypicPlasticity">PrintPhenotypicPlasticity</a><br>
      <a href="#PrintPopulationDistanceData">PrintPopulationDistanceData</a><br>
      <a href="#PrintReactionData">PrintReactionData</a><br>
    <td valign="top">
      <a href="#PrintReactionRewardData">PrintReactionRewardData</a><br>
      <a href="#PrintRelativeFitnessHistogram">PrintRelativeFitnessHistogram</a><br>
      <a href="#print_resource_data">print_resource_data</a><br>
      <a href="#PrintResourceData">PrintResourceData</a><br>
      <a href="#PrintSenseData">PrintSenseData</a><br>
      <a href="#PrintSenseExeData">PrintSenseExeData</a><br>
      <a href="#PrintSleepData">PrintSleepData</a><br>
      <a href="#print_species_abundance_histogram">print_species_abundance_histogram</a><br>
      <a href="#PrintSpeciesAbundanceHistogram">PrintSpeciesAbundanceHistogram</a><br>
      <a href="#print_stats_data">print_stats_data</a><br>
      <a href="#PrintStatsData">PrintStatsData</a><br>
      <a href="#print_tasks_data">print_tasks_data</a><br>
      <a href="#print_tasks_exe_data">print_tasks_exe_data</a><br>
      <a href="#print_tasks_qual_data">print_tasks_qual_data</a><br>
      <a href="#PrintTasksData">PrintTasksData</a><br>
      <a href="#PrintTasksExeData">PrintTasksExeData</a><br>
      <a href="#PrintTaskSnapshot">PrintTaskSnapshot</a><br>
      <a href="#PrintTasksQualData">PrintTasksQualData</a><br>
      <a href="#print_time_data">print_time_data</a><br>
      <a href="#PrintTimeData">PrintTimeData</a><br>
      <a href="#print_totals_data">print_totals_data</a><br>
      <a href="#PrintTotalsData">PrintTotalsData</a><br>
      <a href="#print_tree_depths">print_tree_depths</a><br>
      <a href="#PrintTreeDepths">PrintTreeDepths</a><br>
      <a href="#print_variance_data">print_variance_data</a><br>
      <a href="#PrintVarianceData">PrintVarianceData</a><br>
      <a href="#print_viable_tasks_data">print_viable_tasks_data</a><br>
      <a href="#PrintViableTasksData">PrintViableTasksData</a><br>
      <a href="#RandomLandscape">RandomLandscape</a><br>
      <a href="#rate_kill">rate_kill</a><br>
      <a href="#ReplicateDemes">ReplicateDemes</a><br>
      <a href="#replicate_demes">replicate_demes</a><br>
      <a href="#reset_demes">reset_demes</a><br>
      <a href="#ResetDemes">ResetDemes</a><br>
      <a href="#SampleLandscape">SampleLandscape</a><br>
      <a href="#save_clone">save_clone</a><br>
      <a href="#SaveClone">SaveClone</a><br>
      <a href="#SaveHistoricPopulation">SaveHistoricPopulation</a><br>
      <a href="#SaveHistoricSexPopulation">SaveHistoricSexPopulation</a><br>
      <a href="#SaveParasitePopulation">SaveParasitePopulation</a><br>
      <a href="#SavePopulation">SavePopulation</a><br>
      <a href="#SaveSexPopulation">SaveSexPopulation</a><br>
      <a href="#SerialTransfer">SerialTransfer</a><br>
      <a href="#serial_transfer">serial_transfer</a><br>
      <a href="#SetCellResource">SetCellResource</a><br>
      <a href="#SetEnvironmentInputs">SetEnvironmentInputs</a><br>
      <a href="#SetEnvironmentRandomMask">SetEnvironmentRandomMask</a><br>
      <a href="#SetMutProb">SetMutProb</a><br>
      <a href="#SetNumInstBefore0Energy">SetNumInstBefore0Energy</a><br>
      <a href="#SetOptimizeMinMax">SetOptimizeMinMax</a><br>
      <a href="#SetPeriodicResource">SetPeriodicResource</a><br>
      <a href="#set_reaction_inst">set_reaction_inst</a><br>
      <a href="#set_reaction_value">set_reaction_value</a><br>
      <a href="#set_reaction_value_mult">set_reaction_value_mult</a><br>
      <a href="#SetReactionInst">SetReactionInst</a><br>
      <a href="#SetReactionMaxTaskCount">SetReactionMaxTaskCount</a><br>
      <a href="#SetReactionMinTaskCount">SetReactionMinTaskCount</a><br>
      <a href="#SetReactionTask">SetReactionTask</a><br>
      <a href="#SetReactionValue">SetReactionValue</a><br>
      <a href="#SetReactionValueMult">SetReactionValueMult</a><br>
      <a href="#SetResource">SetResource</a><br>
      <a href="#set_resource">set_resource</a><br>
      <a href="#SetResourceInflow">SetResourceInflow</a><br>
      <a href="#SetResourceOutflow">SetResourceOutflow</a><br>
      <a href="#SetTaskArgDouble">SetTaskArgDouble</a><br>
      <a href="#SetTaskArgInt">SetTaskArgInt</a><br>
      <a href="#SetTaskArgString">SetTaskArgString</a><br>
      <a href="#SetVerbose">SetVerbose</a><br>
      <a href="#sever_grid_col">sever_grid_col</a><br>
      <a href="#sever_grid_row">sever_grid_row</a><br>
      <a href="#SeverGridCol">SeverGridCol</a><br>
      <a href="#SeverGridRow">SeverGridRow</a><br>
      <a href="#StopFastForward">StopFastForward</a><br>
      <a href="#swap_cells">swap_cells</a><br>
      <a href="#SwapCells">SwapCells</a><br>
      <a href="#task_snapshot">task_snapshot</a><br>
      <a href="#test_dom">test_dom</a><br>
      <a href="#TestDominant">TestDominant</a><br>
      <a href="#ToggleFitnessValley">ToggleFitnessValley</a><br>
      <a href="#ToggleRewardInstruction">ToggleRewardInstruction</a><br>
      <a href="#VERBOSE">VERBOSE</a><br>
      <a href="#zero_muts">zero_muts</a><br>
      <a href="#ZeroMuts">ZeroMuts</a><br>
      <a href="#ZeroResources">ZeroResources</a><br>
  </tr>
</table>
<p>&nbsp;</p>
<h2><a name="DriverActions">Driver Actions</a></h2>

<p>
These actions control the driver object responsible for executing the current run.
</p>
<UL>
<li><p>
  <strong><a name="exit">exit</a></strong>
  <i>No Arguments</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="Exit">Exit</a></strong>
  <i>No Arguments</i>
  </p>
  <p>
    Unconditionally terminate the current run.

  </p>
</li>
<li><p>
  <strong><a name="ExitAveGeneration">ExitAveGeneration</a></strong>
  <i>&lt;double generation&gt;</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="ExitAveLineageLabelGreater">ExitAveLineageLabelGreater</a></strong>
  <i>&lt;double threshold&gt;</i>
  </p>
  <p>
     Halts the run if the current average lineage label is larger
   than <span class="cmdarg">threshold</span>.

  </p>
</li>
<li><p>
  <strong><a name="ExitAveLineageLabelLess">ExitAveLineageLabelLess</a></strong>
  <i>&lt;double threshold&gt;</i>
  </p>
  <p>
     Halts the run if the current average lineage label is smaller
   than <span class="cmdarg">threshold</span>.

  </p>
</li>
<li><p>
  <strong><a name="ExitElapsedTime">ExitElapsedTime</a></strong>
  <i>&lt;int elapsed time [seconds]&gt;</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="exit_if_ave_lineage_label_larger">exit_if_ave_lineage_label_larger</a></strong>
  <i>&lt;double threshold&gt;</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="exit_if_ave_lineage_label_smaller">exit_if_ave_lineage_label_smaller</a></strong>
  <i>&lt;double threshold&gt;</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="StopFastForward">StopFastForward</a></strong>
  <i>none</i>
  </p>
  <p>
  
  </p>
</li>
</UL>
<p>&nbsp;</p>
<h2><a name="EnvironmentActions">Environment Actions</a></h2>

<p>
Events that allow user to change environment properties, such as resources
and reaction parameters.
</p>
<UL>
<li><p>
  <strong><a name="ChangeEnvironment">ChangeEnvironment</a></strong>
  <i>&lt;string env_string&gt;</i>
  </p>
  <p>
    Action designed to read in and process a line formatted as if it were in the
  <a href="environment.html">environment</a> file.  This will change 
  environmental settings on the fly.  <b>You should create all resources and 
  reactions in the environment file</b> and only use this file to change these
  resources and reactions. 

  </p>
</li>
<li><p>
  <strong><a name="DelayedDemeEvent">DelayedDemeEvent</a></strong>
  <i>&lt;int x1&gt; &lt;int y1&gt; &lt;int x2&gt; &lt;int y2&gt; &lt;int delay&gt; &lt;int duraion&gt;</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="InjectResource">InjectResource</a></strong>
  <i>&lt;string res_name&gt; &lt;double res_count&gt;</i>
  </p>
  <p>
    Inject (add) a specified amount of a specified resource. 
  <span class="cmdarg">res_name</span> must already exist as
  a resource in environment file.

  </p>
</li>
<li><p>
  <strong><a name="inject_resource">inject_resource</a></strong>
  <i>&lt;string res_name&gt; &lt;double res_count&gt;</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="inject_scaled_resource">inject_scaled_resource</a></strong>
  <i>&lt;string res_name&gt; &lt;double res_count&gt;</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="InjectScaledResource">InjectScaledResource</a></strong>
  <i>&lt;string res_name&gt; &lt;double res_count&gt;</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="outflow_scaled_resource">outflow_scaled_resource</a></strong>
  <i>&lt;string res_name&gt; &lt;double res_percent&gt;</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="OutflowScaledResource">OutflowScaledResource</a></strong>
  <i>&lt;string res_name&gt; &lt;double res_percent&gt;</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="SetCellResource">SetCellResource</a></strong>
  <i>&lt;int cell_id&gt; &lt;string res_name&gt; &lt;double res_count&gt;</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="SetEnvironmentInputs">SetEnvironmentInputs</a></strong>
  <i>&lt;int input_1&gt; &lt;int input_2&gt; &lt;int input_3&gt; </i>
  </p>
  <p>
    Set the inputs that all organisms get from the environment when doing IO to these specific values. There must 
  be exactly three inputs, and they must have the usual values for the top 8 "key" bits, i.e. they must be of the form
  0x0F?????? 0x33?????? 0x55?????? where ? can be replaced with any hexadecimal digit.

  </p>
</li>
<li><p>
  <strong><a name="SetEnvironmentRandomMask">SetEnvironmentRandomMask</a></strong>
  <i>&lt;int mask&gt;</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="SetNumInstBefore0Energy">SetNumInstBefore0Energy</a></strong>
  <i>&lt;int new_value&gt;</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="SetOptimizeMinMax">SetOptimizeMinMax</a></strong>
  <i>No Arguments</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="SetPeriodicResource">SetPeriodicResource</a></strong>
  <i>&lt;string reaction_name&gt; &lt;string amplitude&gt; &lt;string pi/frequence&gt; &lt;phaseShift*pi&gt; &lt;string initial_Y&gt;</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="set_reaction_inst">set_reaction_inst</a></strong>
  <i>&lt;string reaction_name&gt; &lt;string inst&gt;</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="set_reaction_value">set_reaction_value</a></strong>
  <i>&lt;string reaction_name&gt; &lt;double value&gt;</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="set_reaction_value_mult">set_reaction_value_mult</a></strong>
  <i>&lt;string reaction_name&gt; &lt;double value&gt;</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="SetReactionInst">SetReactionInst</a></strong>
  <i>&lt;string reaction_name&gt; &lt;string inst&gt;</i>
  </p>
  <p>
    Set the instruction triggered by this reaction.
  <span class="cmdarg">reaction_name</span> must already
  exist in the environment file.
  <span class="cmdarg">inst</span> must be in the instruction set. 

  </p>
</li>
<li><p>
  <strong><a name="SetReactionMaxTaskCount">SetReactionMaxTaskCount</a></strong>
  <i>&lt;string reaction_name&gt; &lt;int max_count&gt;</i>
  </p>
  <p>
    Set the max task count required to trigger a reaction to <span class="cmdarg">task_count</span>.
  <span class="cmdarg">reaction_name</span> must already 
  exist in the environment file.

  </p>
</li>
<li><p>
  <strong><a name="SetReactionMinTaskCount">SetReactionMinTaskCount</a></strong>
  <i>&lt;string reaction_name&gt; &lt;int min_count&gt;</i>
  </p>
  <p>
    Set the min task count required to trigger a reaction to <span class="cmdarg">task_count</span>.
  <span class="cmdarg">reaction_name</span> must already 
  exist in the environment file.

  </p>
</li>
<li><p>
  <strong><a name="SetReactionTask">SetReactionTask</a></strong>
  <i>&lt;string reaction_name&gt; &lt;string task_name&gt;</i>
  </p>
  <p>
    Set the task required to trigger a reaction to <span class="cmdarg">task_name</span>.
  <span class="cmdarg">reaction_name</span> and <span class="cmdarg">task_name</span> must already 
  exist in the environment file.

  </p>
</li>
<li><p>
  <strong><a name="SetReactionValue">SetReactionValue</a></strong>
  <i>&lt;string reaction_name&gt; &lt;double value&gt;</i>
  </p>
  <p>
    Set the reaction value to a specific level.
  <span class="cmdarg">reaction_name</span> must already 
  exist in the environment file.
  <span class="cmdarg">value</span> can be negative. 

  </p>
</li>
<li><p>
  <strong><a name="SetReactionValueMult">SetReactionValueMult</a></strong>
  <i>&lt;string reaction_name&gt; &lt;double value&gt;</i>
  </p>
  <p>
    Multiply the reaction value by the <span class="cmdarg">value</span>.
  <span class="cmdarg">reaction_name</span> must already 
  exist in the environment file.
  <span class="cmdarg">value</span> can be negative. 

  </p>
</li>
<li><p>
  <strong><a name="SetResource">SetResource</a></strong>
  <i>&lt;string res_name&gt; &lt;double res_count&gt;</i>
  </p>
  <p>
    Set the resource amount to a specific level.
  <span class="cmdarg">res_name</span> must already exist as
  a resource in environment file.

  </p>
</li>
<li><p>
  <strong><a name="set_resource">set_resource</a></strong>
  <i>&lt;string res_name&gt; &lt;double res_count&gt;</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="SetResourceInflow">SetResourceInflow</a></strong>
  <i>&lt;string resource_name&gt; &lt;int inflow&gt;</i>
  </p>
  <p>
    Set the resource inflow to a specific level.
  <span class="cmdarg">res_name</span> must already exist as
  a resource in environment file.

  </p>
</li>
<li><p>
  <strong><a name="SetResourceOutflow">SetResourceOutflow</a></strong>
  <i>&lt;string resource_name&gt; &lt;int outflow&gt;</i>
  </p>
  <p>
    Set the resource outflow to a specific level.
  <span class="cmdarg">res_name</span> must already exist as
  a resource in environment file.

  </p>
</li>
<li><p>
  <strong><a name="SetTaskArgDouble">SetTaskArgDouble</a></strong>
  <i>&lt;int task&gt; &lt;int arg&gt; &lt;double value&gt;</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="SetTaskArgInt">SetTaskArgInt</a></strong>
  <i>&lt;int task&gt; &lt;int arg&gt; &lt;int value&gt;</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="SetTaskArgString">SetTaskArgString</a></strong>
  <i>&lt;int task&gt; &lt;int arg&gt; &lt;string value&gt;</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="ZeroResources">ZeroResources</a></strong>
  <i>none</i>
  </p>
  <p>
    Set all resurce levels to zero.

  </p>
</li>
</UL>
<p>&nbsp;</p>
<h2><a name="LandscapeActions">Landscape Analysis Actions</a></h2>

<p>
Landscape analysis actions perform various types mutation studies to
calculate properties of the fitness landscape for a particular genome.
When scheduled as an event during a run, these actions will typically
perform analysis on the dominant genotype.  In analyze mode, analysis
is performed on the entire currently selected batch.
</p>
<p>
These actions are often very computationally intensive, thus will
take a long time to compute.  In order to take advantage of increasingly
available multi-processor/multi-core systems, a number of these actions
have been enhanced to make use of multiple threads to parallize work.
Set the configuration setting <code>MT_CONCURRENCY</code> to the number
of logical processors available to make use of all processor resources
for these compuations.
</p>
<UL>
<li><p>
  <strong><a name="AnalyzeLandscape">AnalyzeLandscape</a></strong>
  <i>[filename="land-analyze.dat"] [int trials=1000] [int min_found=0] [int max_trials=0] [int max_dist=10]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="AnalyzePopulation">AnalyzePopulation</a></strong>
  <i>[double sample_prob=1] [int landscape=0] [int save_genotype=0] [string filename=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="DeletionLandscape">DeletionLandscape</a></strong>
  <i>[string filename="land-del.dat"] [int distance=1] [string sitecount_file=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="DumpLandscape">DumpLandscape</a></strong>
  <i>[string filename="land-dump.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="FullLandscape">FullLandscape</a></strong>
  <i>[string filename="land-full.dat"] [int distance=1] [string entropy_file=""] [string sitecount_file=""]</i>
  </p>
  <p>
    Do a landscape analysis of the dominant genotype or current batch of genotypes,
  depending on the current mode.  The resulting output is a collection of 
  statistics obtained from examining all possible mutations at the distance
  specified.  The default distance is one.

  </p>
</li>
<li><p>
  <strong><a name="HillClimb">HillClimb</a></strong>
  <i>[string filename="hillclimb.dat"]</i>
  </p>
  <p>
    Does a hill climb with the dominant genotype.

  </p>
</li>
<li><p>
  <strong><a name="InsertionLandscape">InsertionLandscape</a></strong>
  <i>[string filename="land-ins.dat"] [int distance=1] [string sitecount_file=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="MutationalNeighborhood">MutationalNeighborhood</a></strong>
  <i>[string fname="mut-neighborhood.dat"] [int target=-1]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PairTestLandscape">PairTestLandscape</a></strong>
  <i>[string filename=""] [int sample_size=0]</i>
  </p>
  <p>
    If sample_size = 0, pairtest the full landscape.

  </p>
</li>
<li><p>
  <strong><a name="PrecalcLandscape">PrecalcLandscape</a></strong>
  <i>No Arguments</i>
  </p>
  <p>
    Precalculate the distance 1 full landscape for the current batch in parallel
  using multiple threads.  The resulting data is stored into the current batch
  and can be used by many subsequent output commands within Analyze mode.

  </p>
</li>
<li><p>
  <strong><a name="PredictNuLandscape">PredictNuLandscape</a></strong>
  <i>[string filename="land-predict.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PredictWLandscape">PredictWLandscape</a></strong>
  <i>[string filename="land-predict.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="RandomLandscape">RandomLandscape</a></strong>
  <i>[string filename="land-random.dat"] [int distance=1] [int trials=0]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="SampleLandscape">SampleLandscape</a></strong>
  <i>[string filename="land-sample.dat"] [int trials=0]</i>
  </p>
  <p>
  
  </p>
</li>
</UL>
<p>&nbsp;</p>
<h2><a name="PopulationActions">Population Actions</a></h2>

<p>
Population events modify the state of the population, and will actually
change the course of the run.  There are a wide variety of these.
</p>
<UL>
<li><p>
  <strong><a name="apocalypse">apocalypse</a></strong>
  <i>[double probability=0.9]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="compete_demes">compete_demes</a></strong>
  <i>[int type=1]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="CompeteDemes">CompeteDemes</a></strong>
  <i>[int type=1]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="compete_organisms">compete_organisms</a></strong>
  <i>[int type=1]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="CompeteOrganisms">CompeteOrganisms</a></strong>
  <i>[int type=0] [int parents_survive=0] [double scaled_time=1.0] [int dynamic_scaling=0]</i>
  </p>
  <p>
    Calculates fitness of each organism in the population and creates a new population of descendants
  where each organism has a chance of replication proportional to its fitness. Can be used to make Avida 
  operate without population dynamics and asynchronous replication, i.e. like a genetic algorithm. If NewTrial 
  events have occurred since the last CompeteOrganisms event, then the average fitness over those trials 
  will be used.
  </p>
 
  <code>competition_type</code> controls how the fitnesses of multiple trials determine the overall fitness
  used for the competition: 0=geometric mean of fitnesses, 1=scaled geometric mean of fitnesses (the greatest fitness
  of each trial is scaled to 1.0 before taking the geometric mean), 2=arithmetic mean, 3=geometric mean plus 
  rescales effective fitness values by the geometric mean of the difference from the top score and the median.
 <p>Setting <code>parents_survive</code> to 1, causes the first copy of an organism that makes it into the new 
  population to be the original (unmutated) parent organism.

  </p>
</li>
<li><p>
  <strong><a name="connect_cells">connect_cells</a></strong>
  <i>&lt;int cellA_x&gt; &lt;int cellA_y&gt; &lt;int cellB_x&gt; &lt;int cellB_y&gt;</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="ConnectCells">ConnectCells</a></strong>
  <i>&lt;int cellA_x&gt; &lt;int cellA_y&gt; &lt;int cellB_x&gt; &lt;int cellB_y&gt;</i>
  </p>
  <p>
    Connects a pair of specified cells.

  </p>
</li>
<li><p>
  <strong><a name="CopyDeme">CopyDeme</a></strong>
  <i>&lt;int src_id&gt; &lt;int dest_id&gt;</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="copy_deme">copy_deme</a></strong>
  <i>&lt;int src_id&gt; &lt;int dest_id&gt;</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="DisconnectCells">DisconnectCells</a></strong>
  <i>&lt;int cellA_x&gt; &lt;int cellA_y&gt; &lt;int cellB_x&gt; &lt;int cellB_y&gt;</i>
  </p>
  <p>
    Disconnects a pair of specified cells.

  </p>
</li>
<li><p>
  <strong><a name="disconnect_cells">disconnect_cells</a></strong>
  <i>&lt;int cellA_x&gt; &lt;int cellA_y&gt; &lt;int cellB_x&gt; &lt;int cellB_y&gt;</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="DivideDemes">DivideDemes</a></strong>
  <i>No arguments (yet!)</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="inject">inject</a></strong>
  <i>[string fname="START_CREATURE"] [int cell_id=0] [double merit=-1] [int lineage_label=0] [double neutral_metric=0]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="Inject">Inject</a></strong>
  <i>[string fname="START_CREATURE"] [int cell_id=0] [double merit=-1] [int lineage_label=0] [double neutral_metric=0]</i>
  </p>
  <p>
    Inject a single organisms into the population.  Arguments must be
  included from left to right; if all arguments are left out, the default
  creature is the ancestral organism, and it will be injected into cell 0,
  have an uninitialized merit, and be marked as liniage id 0.

  </p>
</li>
<li><p>
  <strong><a name="inject_all">inject_all</a></strong>
  <i>[string fname="START_CREATURE"] [double merit=-1] [int lineage_label=0] [double neutral_metric=0]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="InjectAll">InjectAll</a></strong>
  <i>[string fname="START_CREATURE"] [double merit=-1] [int lineage_label=0] [double neutral_metric=0]</i>
  </p>
  <p>
    Same as Inject, but no cell_id is specified and the organism is placed
  into <em>all</em> cells in the population.

  </p>
</li>
<li><p>
  <strong><a name="inject_all_random_repro">inject_all_random_repro</a></strong>
  <i>&lt;int length&gt; [double merit=-1] [int lineage_label=0] [double neutral_metric=0]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="InjectDemes">InjectDemes</a></strong>
  <i>[string fname="START_CREATURE"] [double merit=-1] [int lineage_label=0] [double neutral_metric=0]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="InjectParasite">InjectParasite</a></strong>
  <i>&lt;string filename&gt; &lt;string label&gt; [int cell_start=0] [int cell_end=-1]</i>
  </p>
  <p>
    Attempt to inject a parasite genome into the supplied population cell
  range with the specified label.

  </p>
</li>
<li><p>
  <strong><a name="InjectParasitePair">InjectParasitePair</a></strong>
  <i>&lt;string filename_genome&gt; &lt;string filename_parasite&gt; &lt;string label&gt; [int cell_start=0] [int cell_end=-1] [double merit=-1] [int lineage_label=0] [double neutral_metric=0]</i>
  </p>
  <p>
    Inject host parasite pairs into the population cell range specified. 

  </p>
</li>
<li><p>
  <strong><a name="inject_random">inject_random</a></strong>
  <i>&lt;int length&gt; [int cell_id=0] [double merit=-1] [int lineage_label=0] [double neutral_metric=0]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="InjectRandom">InjectRandom</a></strong>
  <i>&lt;int length&gt; [int cell_id=0] [double merit=-1] [int lineage_label=0] [double neutral_metric=0]</i>
  </p>
  <p>
    Injects a randomly generated genome of the supplied length into the population.

  </p>
</li>
<li><p>
  <strong><a name="InjectRange">InjectRange</a></strong>
  <i>[string fname="START_CREATURE"] [int cell_start=0] [int cell_end=-1] [double merit=-1] [int lineage_label=0] [double neutral_metric=0]</i>
  </p>
  <p>
    Injects identical organisms into a range of cells of the population.
  </p>
  <p>
  <em>Example</em>:
  <br /><code>InjectRange 000-aaaaa.org 0 10</code>
  <br />Will inject 10 organisms into cells 0 through 9.

  </p>
</li>
<li><p>
  <strong><a name="inject_range">inject_range</a></strong>
  <i>[string fname="START_CREATURE"] [int cell_start=0] [int cell_end=-1] [double merit=-1] [int lineage_label=0] [double neutral_metric=0]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="inject_sequence">inject_sequence</a></strong>
  <i>[string fname="START_CREATURE"] [int cell_id=0] [double merit=-1] [int lineage_label=0] [double neutral_metric=0]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="InjectSequence">InjectSequence</a></strong>
  <i>&lt;string sequence&gt; [int cell_start=0] [int cell_end=-1] [double merit=-1] [int lineage_label=0] [double neutral_metric=0]</i>
  </p>
  <p>
    Injects identical organisms based on the supplied genome sequence into
  a range of cells of the population.
  </p>
  <p>
  <em>Example</em>:
  <br /><code>InjectSequence ckdfhgklsahnfsaggdsgajfg 0 10 100</code>
  <br />Will inject 10 organisms into cells 0 through 9 with a merit of 100.

  </p>
</li>
<li><p>
  <strong><a name="InjectSequenceWDivMutRate">InjectSequenceWDivMutRate</a></strong>
  <i>&lt;string sequence&gt; [int cell_start=0] [int cell_end=-1] [double div_mut_rate=0] [double merit=-1] [int lineage_label=0] [double neutral_metric=0]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="join_grid_col">join_grid_col</a></strong>
  <i>[int col_id=-1] [int min_row=0] [int max_row=-1]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="join_grid_row">join_grid_row</a></strong>
  <i>[int row_id=-1] [int min_col=0] [int max_col=-1]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="JoinGridCol">JoinGridCol</a></strong>
  <i>[int col_id=-1] [int min_row=0] [int max_row=-1]</i>
  </p>
  <p>
    Add connections between cells along a column in an Avida grid.

  </p>
</li>
<li><p>
  <strong><a name="JoinGridRow">JoinGridRow</a></strong>
  <i>[int row_id=-1] [int min_col=0] [int max_col=-1]</i>
  </p>
  <p>
    Add connections between cells along a row in an Avida grid.

  </p>
</li>
<li><p>
  <strong><a name="KillProb">KillProb</a></strong>
  <i>[double probability=0.9]</i>
  </p>
  <p>
    Using the specified probability, test each organism to see if it is killed off.

  </p>
</li>
<li><p>
  <strong><a name="KillRate">KillRate</a></strong>
  <i>[double probability=0.9]</i>
  </p>
  <p>
    Randomly removes a certain proportion of the population.
  In principle, this action does the same thing as the KillProb event.
  However, instead of a probability, here one has to specify a rate. The
  rate has the same unit as fitness. So if the average fitness is 20000,
  than you remove 50% of the population on every update with a removal rate
  of 10000.

  </p>
</li>
<li><p>
  <strong><a name="KillRectangle">KillRectangle</a></strong>
  <i>[int x1=0] [int y1=0] [int x2=0] [int y2=0]</i>
  </p>
  <p>
    Kill off all organisms in a rectangle defined by the points (x1, y1) and (x2, y2).

  </p>
</li>
<li><p>
  <strong><a name="kill_rectangle">kill_rectangle</a></strong>
  <i>[int x1=0] [int y1=0] [int x2=0] [int y2=0]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="ModMutProb">ModMutProb</a></strong>
  <i>[string mut_type="copy"] [double prob=0.0] [int start_cell=-1] [int end_cell=-1]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="NewTrial">NewTrial</a></strong>
  <i>No Arguments</i>
  </p>
  <p>
    Immediately calculates the fitness of each organism in the population, saves this value for use with the CompeteOrganisms
  event, and resets the state of all organisms.

  </p>
</li>
<li><p>
  <strong><a name="new_trial">new_trial</a></strong>
  <i>[int type=1]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="rate_kill">rate_kill</a></strong>
  <i>&lt;double probability&gt;</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="ReplicateDemes">ReplicateDemes</a></strong>
  <i>[string trigger=full_deme]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="replicate_demes">replicate_demes</a></strong>
  <i>[string trigger=full_deme]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="reset_demes">reset_demes</a></strong>
  <i>No Arguments</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="ResetDemes">ResetDemes</a></strong>
  <i>No Arguments</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="SerialTransfer">SerialTransfer</a></strong>
  <i>[int transfer_size=1] [int ignore_deads=1]</i>
  </p>
  <p>
    Similar to KillProb, but we specify the exact number of organisms to
  keep alive after the event.  The <span class="cmdarg">ignore_deads</span>
  argument determines whether only living organisms are retainted.

  </p>
</li>
<li><p>
  <strong><a name="serial_transfer">serial_transfer</a></strong>
  <i>[int transfer_size=1] [int ignore_deads=1]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="SetMutProb">SetMutProb</a></strong>
  <i>[string mut_type="copy"] [double prob=0.0] [int start_cell=-1] [int end_cell=-1]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="sever_grid_col">sever_grid_col</a></strong>
  <i>[int col_id=-1] [int min_row=0] [int max_row=-1]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="sever_grid_row">sever_grid_row</a></strong>
  <i>[int row_id=-1] [int min_col=0] [int max_col=-1]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="SeverGridCol">SeverGridCol</a></strong>
  <i>[int col_id=-1] [int min_row=0] [int max_row=-1]</i>
  </p>
  <p>
    Remove the connections between cells along a column in an Avida grid.

  </p>
</li>
<li><p>
  <strong><a name="SeverGridRow">SeverGridRow</a></strong>
  <i>[int row_id=-1] [int min_col=0] [int max_col=-1]</i>
  </p>
  <p>
    Remove the connections between cells along a row in an Avida grid.

  </p>
</li>
<li><p>
  <strong><a name="swap_cells">swap_cells</a></strong>
  <i>&lt;int cell_id1&gt; &lt;int cell_id2&gt;</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="SwapCells">SwapCells</a></strong>
  <i>&lt;int cell_id1&gt; &lt;int cell_id2&gt;</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="ToggleFitnessValley">ToggleFitnessValley</a></strong>
  <i>No Arguments</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="ToggleRewardInstruction">ToggleRewardInstruction</a></strong>
  <i>No Arguments</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="zero_muts">zero_muts</a></strong>
  <i>No Arguments</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="ZeroMuts">ZeroMuts</a></strong>
  <i>No Arguments</i>
  </p>
  <p>
    This event will set all mutation rates to zero.

  </p>
</li>
</UL>
<p>&nbsp;</p>
<h2><a name="PrintActions">Print Actions</a></h2>

<p>
Output events are the primary way of saving data from an Avida experiments.
The main two types are <em>continuous output</em>, which append to a single file 
every time the event is trigged, and <em>singular output</em>, which produce
a single, complete file for each trigger.
</p>
<UL>
<li><p>
  <strong><a name="DumpCellDataGrid">DumpCellDataGrid</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="dump_donor_grid">dump_donor_grid</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="DumpDonorGrid">DumpDonorGrid</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
    Print out the grid of organisms who donated their merit.

  </p>
</li>
<li><p>
  <strong><a name="DumpEnergyGrid">DumpEnergyGrid</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="DumpExecutionRatioGrid">DumpExecutionRatioGrid</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="dump_fitness_grid">dump_fitness_grid</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="DumpFitnessGrid">DumpFitnessGrid</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
    Print out the grid of organism fitness values.

  </p>
</li>
<li><p>
  <strong><a name="dump_genotype_grid">dump_genotype_grid</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="DumpGenotypeIDGrid">DumpGenotypeIDGrid</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
    Print out the grid of genotype IDs.

  </p>
</li>
<li><p>
  <strong><a name="dump_lineage_grid">dump_lineage_grid</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="DumpLineageGrid">DumpLineageGrid</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="dump_memory">dump_memory</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="DumpMemory">DumpMemory</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
    Dump memory summary information.

  </p>
</li>
<li><p>
  <strong><a name="dump_phenotype_grid">dump_phenotype_grid</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="DumpPhenotypeIDGrid">DumpPhenotypeIDGrid</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="dump_receiver_grid">dump_receiver_grid</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="DumpReceiverGrid">DumpReceiverGrid</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="DumpSleepGrid">DumpSleepGrid</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="dump_task_grid">dump_task_grid</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="DumpTaskGrid">DumpTaskGrid</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
    Print out the grid of takss that organisms do. For each organism, tasks are first 
  encoded as a binary string (e.g. 100000001 means that organism is doing NOT and EQU
  and then reported as a base-10 number (257 in the example above).

  </p>
</li>
<li><p>
  <strong><a name="Echo">Echo</a></strong>
  <i>&lt;cString message&gt;</i>
  </p>
  <p>
    Print the supplied message to standard output.

  </p>
</li>
<li><p>
  <strong><a name="echo">echo</a></strong>
  <i>&lt;cString message&gt;</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="genetic_distance_pop_dump">genetic_distance_pop_dump</a></strong>
  <i>[string creature="START_CREATURE"] [string fname=""] [int save_genotypes=0]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="print_average_data">print_average_data</a></strong>
  <i>[string fname="average.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintAverageData">PrintAverageData</a></strong>
  <i>[string fname="average.dat"]</i>
  </p>
  <p>
    [<span class="cmdarg">string filename='average.dat'</span>]
  Print all of the population averages the specified file.

  </p>
</li>
<li><p>
  <strong><a name="PrintCCladeCounts">PrintCCladeCounts</a></strong>
  <i>[filename = "cclade_count.dat"]</i>
  </p>
  <p>
      Print a count of the number of oraganisms belonging to a each coalescence clade currently in the population.
    <br />This action will only work in run mode.
    <br />This action will require TRACK_CCLADE to be enabled.

  </p>
</li>
<li><p>
  <strong><a name="PrintCCladeFitnessHistogram">PrintCCladeFitnessHistogram</a></strong>
  <i>[filename] [fit_mode] [hist_min] [hist_step] [hist_max]</i>
  </p>
  <p>
      Print a histogram of fitnesses for each coalescence clade in the population.
    <br />This action will only work in run mode.
    <br />This action will rerequire TRACK_CCLADE to be enabled.
    <br/>
    <kbd>mode</kbd> may be {<kbd>CURRENT, ACTUAL, TESTCPU</kbd>}, where
      <br/>&nbsp; &nbsp; &nbsp;<kbd>CURRENT</kbd> uses the current phenotype value of fitness</li>
      <br/>&nbsp; &nbsp; &nbsp;<kbd>ACTUAL</kbd> uses the current merit and the true gestation time (via test cpu calculation)</li>
      <br/>&nbsp; &nbsp; &nbsp;<kbd>TESTCPU</kbd> uses the test cpu measurement.</li>
      <br />
    <kbd>lower_bound, step, upper_bound</kbd> are log10 values for the individual histogram bins.</kbd>

  </p>
</li>
<li><p>
  <strong><a name="PrintCCladeRelativeFitnessHistogram">PrintCCladeRelativeFitnessHistogram</a></strong>
  <i>[filename] [fit_mode] [hist_min] [hist_step] [hist_max]</i>
  </p>
  <p>
      Print a histogram of parent-relative fitness ratios for each coalescence clade in the population.
    <br />This action will only work in run mode.
    <br />This action will rerequire TRACK_CCLADE to be enabled.
    <br/>
    <kbd>mode</kbd> may be {<kbd>CURRENT, ACTUAL, TESTCPU</kbd>}, where
      <br/>&nbsp; &nbsp; &nbsp;<kbd>CURRENT</kbd> uses the current phenotype value of fitness</li>
      <br/>&nbsp; &nbsp; &nbsp;<kbd>ACTUAL</kbd> uses the current merit and the true gestation time (via test cpu calculation)</li>
      <br/>&nbsp; &nbsp; &nbsp;<kbd>TESTCPU</kbd> uses the test cpu measurement.</li>
      <br />
    <kbd>lower_bound, step, upper_bound</kbd> are values for the individual histogram bins.</kbd>

  </p>
</li>
<li><p>
  <strong><a name="PrintCellVisitsData">PrintCellVisitsData</a></strong>
  <i>[string fname="visits.dat			"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintCompetitionData">PrintCompetitionData</a></strong>
  <i>[string fname="competition.dat"]</i>
  </p>
  <p>
    Print out CompeteOrganism statistics. Make the first call after the first set of trials has been completed to get a complete header.

  </p>
</li>
<li><p>
  <strong><a name="print_count_data">print_count_data</a></strong>
  <i>[string fname="count.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintCountData">PrintCountData</a></strong>
  <i>[string fname="count.dat"]</i>
  </p>
  <p>
    Print all of the statistics the keep track of counts (such as the number of organisms
  in the population or the number of instructions executed).

  </p>
</li>
<li><p>
  <strong><a name="PrintCurrentReactionData">PrintCurrentReactionData</a></strong>
  <i>[string fname="cur_reactions.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="print_data">print_data</a></strong>
  <i>&lt;cString fname&gt; &lt;cString format&gt;</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintData">PrintData</a></strong>
  <i>&lt;cString fname&gt; &lt;cString format&gt;</i>
  </p>
  <p>
    Append to the file specified (continuous output), the data given in the
  column list.  The column list needs to be a comma-seperated list of
  keywords representing the data types.  Many possible data types can be
  output; see the <a href="print_data.html">complete listing</a> for details.
  Note that this event will even create a detailed column legend at the top
  of your file so you don't need to seperately keep track of what the
  columns mean.

  </p>
</li>
<li><p>
  <strong><a name="PrintDebug">PrintDebug</a></strong>
  <i>No Arguments</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="print_deme_stats">print_deme_stats</a></strong>
  <i>No Arguments</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintDemeAllStats">PrintDemeAllStats</a></strong>
  <i>No Arguments</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintDemeAverageData">PrintDemeAverageData</a></strong>
  <i>[string fname="deme_average.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintDemeDonorStats">PrintDemeDonorStats</a></strong>
  <i>No Arguments</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintDemeReplicationData">PrintDemeReplicationData</a></strong>
  <i>[string fname="deme_repl.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintDemeResourceStats">PrintDemeResourceStats</a></strong>
  <i>No Arguments</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintDemeSpacialEnergyStats">PrintDemeSpacialEnergyStats</a></strong>
  <i>No Arguments</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintDemeSpacialSleepStats">PrintDemeSpacialSleepStats</a></strong>
  <i>No Arguments</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintDemeStats">PrintDemeStats</a></strong>
  <i>No Arguments</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="print_depth_histogram">print_depth_histogram</a></strong>
  <i>[string fname="depth_histogram.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintDepthHistogram">PrintDepthHistogram</a></strong>
  <i>[string fname="depth_histogram.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="print_detailed_fitness_data">print_detailed_fitness_data</a></strong>
  <i>[int save_max_f_genotype=0] [int print_fitness_histo=0] [double hist_fmax=1] [double hist_fstep=0.1] [string datafn="fitness.dat"] [string histofn="fitness_histos.dat"] [string histotestfn="fitness_histos_testCPU.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintDetailedFitnessData">PrintDetailedFitnessData</a></strong>
  <i>[int save_max_f_genotype=0] [int print_fitness_histo=0] [double hist_fmax=1] [double hist_fstep=0.1] [string datafn="fitness.dat"] [string histofn="fitness_histos.dat"] [string histotestfn="fitness_histos_testCPU.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="print_divide_mut_data">print_divide_mut_data</a></strong>
  <i>[string fname="divide_mut.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintDivideMutData">PrintDivideMutData</a></strong>
  <i>[string fname="divide_mut.dat"]</i>
  </p>
  <p>
    Output (regular and log) statistics about individual, per site, rates divide mutation
  rates (aver, stdev, skew, cur) to divide_mut.dat. Use with multiple divide instuction set.

  </p>
</li>
<li><p>
  <strong><a name="print_dom">print_dom</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="print_dom_parasite">print_dom_parasite</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="print_dominant_data">print_dominant_data</a></strong>
  <i>[string fname="dominant.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintDominantData">PrintDominantData</a></strong>
  <i>[string fname="dominant.dat"]</i>
  </p>
  <p>
    Print all of the statistics relating to the dominant genotype.

  </p>
</li>
<li><p>
  <strong><a name="PrintDominantGenotype">PrintDominantGenotype</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
    Print the dominant organism's genome (and lots of information about it)
  into the file specified.  If no filename is given, the genotype's assigned name
  is used and the file is placed into the archive subdirectory.

  </p>
</li>
<li><p>
  <strong><a name="PrintDominantParasiteGenotype">PrintDominantParasiteGenotype</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
    Print the dominant parasite's genome (and lots of information about it)
  into the file specified.  If no filename is given, the parasite's assigned name
  is used and the file is placed into the archive subdirectory.

  </p>
</li>
<li><p>
  <strong><a name="print_donation_stats">print_donation_stats</a></strong>
  <i>No Arguments</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="print_error_data">print_error_data</a></strong>
  <i>[string fname="error.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintErrorData">PrintErrorData</a></strong>
  <i>[string fname="error.dat"]</i>
  </p>
  <p>
    Print all of the standard errors of the average population statistics.

  </p>
</li>
<li><p>
  <strong><a name="print_genetic_distance_data">print_genetic_distance_data</a></strong>
  <i>[string ref_creature_file="START_CREATURE"] [string fname="genetic_distance.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintGeneticDistanceData">PrintGeneticDistanceData</a></strong>
  <i>[string ref_creature_file="START_CREATURE"] [string fname="genetic_distance.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintGenomicSiteEntropy">PrintGenomicSiteEntropy</a></strong>
  <i>[filename = "GenomicSiteEntropyData.datcd "]</i>
  </p>
  <p>
     This function will take the initial genotype for each organism in the
   population/batch, align them, and calculate the per-site entropy of the
   aligned sequences.  Please note that there may be a variable number
   of columns in each line if the runs are not fixed length.  The site
   entropy will be measured in mers, normalized by the instruction set size.
   This is a population/batch measure of entropy, not a mutation-selection balance
   measure.  

  </p>
</li>
<li><p>
  <strong><a name="print_genotype_abundance_histogram">print_genotype_abundance_histogram</a></strong>
  <i>[string fname="genotype_abundance_histogram.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="print_genotype_map">print_genotype_map</a></strong>
  <i>[string fname="genotype_map.m"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintGenotypeAbundanceHistogram">PrintGenotypeAbundanceHistogram</a></strong>
  <i>[string fname="genotype_abundance_histogram.dat"]</i>
  </p>
  <p>
    Writes out a genotype abundance histogram.

  </p>
</li>
<li><p>
  <strong><a name="PrintGenotypeMap">PrintGenotypeMap</a></strong>
  <i>[string fname="genotype_map.m"]</i>
  </p>
  <p>
    This event is used to output a map of the genotype IDs for the
  population grid to a file that is suitable to be read into Matlab.

  </p>
</li>
<li><p>
  <strong><a name="print_genotypes">print_genotypes</a></strong>
  <i>[string data_fields="all"] [int historic=0] [string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintGenotypes">PrintGenotypes</a></strong>
  <i>[string data_fields="all"] [int historic=0] [string fname=""]</i>
  </p>
  <p>
    This command is used to print out information about all of the genotypes
  in the population.  The file output from here can be read
  back into the analyze mode of Avida with the <code>LOAD</code> command.
  </p>
  <p>
  The <span class="cmdarg">data_fields</span> parameter indicates what
  columns should be included in the file, which must be comma seperated.
  Options are:
  <em>all, id, parent_id, parent2_id (for sex), parent_dist, num_cpus,
  total_cpus, length, merit, gest_time, fitness, update_born, update_dead,
  depth, lineage, sequence.</em>
  Use <em>all</em> (the default) if you want all of the fields included.
  </p>
  <p>
  The <span class="cmdarg">print_historic</span> parameter indicates how
  many updates back in time should be included in this output.  For example,
  '200' would indicate that any ancestor of the current population that died
  out in the last 200 updates should also be printed. A '-1' in this field
  indicates that all ancestors should be printed.
  </p>
  <p>
  The <span class="cmdarg">filename</span> parameter simply indicates what
  you want to call the file.
  </p>
  <p>
  <em>Example</em>:
  <br />&nbsp;&nbsp;<kbd>u 1000:1000 print_genotypes id,parent_id,fitness 1000</kbd>
  <br />This will print out the full population every 1000 updates, including all
  genotypes that have died out since the last time it was printed.

  </p>
</li>
<li><p>
  <strong><a name="PrintGermlineData">PrintGermlineData</a></strong>
  <i>[string fname="germline.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="print_instruction_abundance_histogram">print_instruction_abundance_histogram</a></strong>
  <i>[string fname="instruction_histogram.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="print_instruction_data">print_instruction_data</a></strong>
  <i>[string fname="instruction.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintInstructionAbundanceHistogram">PrintInstructionAbundanceHistogram</a></strong>
  <i>[string fname="instruction_histogram.dat"]</i>
  </p>
  <p>
    Appends a line containing the bulk count (abundance) of each instruction
  in the population onto a file.

  </p>
</li>
<li><p>
  <strong><a name="PrintInstructionData">PrintInstructionData</a></strong>
  <i>[string fname="instruction.dat"]</i>
  </p>
  <p>
    Print the by-organisms counts of what instructions they _successfully_ executed
  beteween birth and divide. Prior to their first divide, organisms values for their parents.

  </p>
</li>
<li><p>
  <strong><a name="print_lineage_counts">print_lineage_counts</a></strong>
  <i>[string fname="lineage_counts.dat"] [int verbose=1]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="print_lineage_totals">print_lineage_totals</a></strong>
  <i>[string fname="lineage_totals.dat"] [int verbose=1]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintLineageCounts">PrintLineageCounts</a></strong>
  <i>[string fname="lineage_counts.dat"] [int verbose=1]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintLineageTotals">PrintLineageTotals</a></strong>
  <i>[string fname="lineage_totals.dat"] [int verbose=1]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintLogFitnessHistogram">PrintLogFitnessHistogram</a></strong>
  <i>Parameters: &lt;filename&gt; &lt;mode&gt; &lt;min&gt; &lt;step&gt; &lt;max&gt;</i>
  </p>
  <p>
      Print a histogram of organism fitnesses in the current population.
    <br />This action will only work in run mode.
    <br/>
    <kbd>mode</kbd> may be {<kbd>CURRENT, ACTUAL, TESTCPU</kbd>}, where
      <br/>&nbsp; &nbsp; &nbsp;<kbd>CURRENT</kbd> uses the current phenotype value of fitness</li>
      <br/>&nbsp; &nbsp; &nbsp;<kbd>ACTUAL</kbd> uses the current merit and the true gestation time (via test cpu calculation)</li>
      <br/>&nbsp; &nbsp; &nbsp;<kbd>TESTCPU</kbd> uses the test cpu measurement.</li>
      <br />
    <kbd>lower_bound, step, upper_bound</kbd> are log10 values for the individual histogram bins.</kbd>

  </p>
</li>
<li><p>
  <strong><a name="print_market_data">print_market_data</a></strong>
  <i>[string fname="market.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintMarketData">PrintMarketData</a></strong>
  <i>[string fname="market.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="print_mutation_rate_data">print_mutation_rate_data</a></strong>
  <i>[string fname="mutation_rates.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintMutationRateData">PrintMutationRateData</a></strong>
  <i>[string fname="mutation_rates.dat"]</i>
  </p>
  <p>
    Output (regular and log) statistics about individual copy mutation rates
  (aver, stdev, skew, cur). Useful only when mutation rate is set per organism.

  </p>
</li>
<li><p>
  <strong><a name="print_number_phenotypes">print_number_phenotypes</a></strong>
  <i>[string fname="phenotype_count.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintParasiteData">PrintParasiteData</a></strong>
  <i>[string fname="parasite.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="print_phenotype_status">print_phenotype_status</a></strong>
  <i>[string fname="phenotype_status.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintPhenotypeData">PrintPhenotypeData</a></strong>
  <i>[string fname="phenotype_count.dat"]</i>
  </p>
  <p>
    [<span class="cmdarg">string filename='phenotype_count.dat'</span>]
  Print the number of phenotypes based on tasks executed this update.  Executing
  a task any number of times is considered the same as executing it once. 

  </p>
</li>
<li><p>
  <strong><a name="PrintPhenotypeStatus">PrintPhenotypeStatus</a></strong>
  <i>[string fname="phenotype_status.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintPhenotypicPlasticity">PrintPhenotypicPlasticity</a></strong>
  <i>[string filename="phenplast"] [int num_trials=1000]</i>
  </p>
  <p>
    This function will provided detailed information about the phenotypic varients
  of the current population/batch by running each genome through a test cpu <kbd>num_trials</kbd> times.
  If this command is executed in run mode, the
  <kbd>filename</kbd> will be appeneded with <kbd>-Update.dat</kbd> where <kbd>Update</kbd>
  is the current update.  In analyze mode, the default file is merely </kbd>phenplast.dat</kbd>.
  The output file contains the following: id, parent_id, phenotypic_varient_number, frequency, fitness, merit, 
  gestation_time, and task counts for each phenotypic variant of each genotype.

  </p>
</li>
<li><p>
  <strong><a name="PrintPopulationDistanceData">PrintPopulationDistanceData</a></strong>
  <i>[string creature="START_CREATURE"] [string fname=""] [int save_genotypes=0]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintReactionData">PrintReactionData</a></strong>
  <i>[string fname="reactions.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintReactionRewardData">PrintReactionRewardData</a></strong>
  <i>[string fname="reaction_reward.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintRelativeFitnessHistogram">PrintRelativeFitnessHistogram</a></strong>
  <i>[filename] [fit_mode] [hist_min] [hist_step] [hist_max]</i>
  </p>
  <p>
      Print a histogram of parent-relative fitness ratios for each coalescence clade in the population.
    <br />This action will only work in run mode.
    <br/>
    <kbd>mode</kbd> may be {<kbd>CURRENT, ACTUAL, TESTCPU</kbd>}, where
      <br/>&nbsp; &nbsp; &nbsp;<kbd>CURRENT</kbd> uses the current phenotype value of fitness</li>
      <br/>&nbsp; &nbsp; &nbsp;<kbd>ACTUAL</kbd> uses the current merit and the true gestation time (via test cpu calculation)</li>
      <br/>&nbsp; &nbsp; &nbsp;<kbd>TESTCPU</kbd> uses the test cpu measurement.</li>
      <br />
    <kbd>lower_bound, step, upper_bound</kbd> are values for the individual histogram bins.</kbd>

  </p>
</li>
<li><p>
  <strong><a name="print_resource_data">print_resource_data</a></strong>
  <i>[string fname="resource.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintResourceData">PrintResourceData</a></strong>
  <i>[string fname="resource.dat"]</i>
  </p>
  <p>
    Print the current counts of each resource available to the population.  This uses
  the environment configuration to determine what resources are in use.  Also creates
  seperate files <kbd>resource_<em>resource_name</em>.m</kbd> (in a format that is
  designed to be read into Matlab) for each spatial resource.

  </p>
</li>
<li><p>
  <strong><a name="PrintSenseData">PrintSenseData</a></strong>
  <i>[string fname="sense.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintSenseExeData">PrintSenseExeData</a></strong>
  <i>[string fname="sense_exe.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintSleepData">PrintSleepData</a></strong>
  <i>[string fname="sleep.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="print_species_abundance_histogram">print_species_abundance_histogram</a></strong>
  <i>[string fname="species_abundance_histogram.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintSpeciesAbundanceHistogram">PrintSpeciesAbundanceHistogram</a></strong>
  <i>[string fname="species_abundance_histogram.dat"]</i>
  </p>
  <p>
    Writes out a species abundance histogram.

  </p>
</li>
<li><p>
  <strong><a name="print_stats_data">print_stats_data</a></strong>
  <i>[string fname="stats.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintStatsData">PrintStatsData</a></strong>
  <i>[string fname="stats.dat"]</i>
  </p>
  <p>
    Print all of the miscellanous population statistics.

  </p>
</li>
<li><p>
  <strong><a name="print_tasks_data">print_tasks_data</a></strong>
  <i>[string fname="tasks.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="print_tasks_exe_data">print_tasks_exe_data</a></strong>
  <i>[string fname="tasks_exe.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="print_tasks_qual_data">print_tasks_qual_data</a></strong>
  <i>[string fname="tasks_quality.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintTasksData">PrintTasksData</a></strong>
  <i>[string fname="tasks.dat"]</i>
  </p>
  <p>
    Print the number of organisms that are able to perform each task.  This uses the
  environment configuration to determine what tasks are in use.

  </p>
</li>
<li><p>
  <strong><a name="PrintTasksExeData">PrintTasksExeData</a></strong>
  <i>[string fname="tasks_exe.dat"]</i>
  </p>
  <p>
    Print number of times the particular task has been executed this update.

  </p>
</li>
<li><p>
  <strong><a name="PrintTaskSnapshot">PrintTaskSnapshot</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
    Run all organisms in the population through test cpus and print out the
  number of tasks each can perform.

  </p>
</li>
<li><p>
  <strong><a name="PrintTasksQualData">PrintTasksQualData</a></strong>
  <i>[string fname="tasks_quality.dat"]</i>
  </p>
  <p>
    Print the total quality of each task.  By default a successful task is valued
  as 1.0.  Some tasks, however, can grant partial values and/or special bonuses
  via the quality value.

  </p>
</li>
<li><p>
  <strong><a name="print_time_data">print_time_data</a></strong>
  <i>[string fname="time.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintTimeData">PrintTimeData</a></strong>
  <i>[string fname="time.dat"]</i>
  </p>
  <p>
    Print all of the timing related statistics.

  </p>
</li>
<li><p>
  <strong><a name="print_totals_data">print_totals_data</a></strong>
  <i>[string fname="totals.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintTotalsData">PrintTotalsData</a></strong>
  <i>[string fname="totals.dat"]</i>
  </p>
  <p>
    Print various totals for the entire length of the run (for example, the total number of
  organisms ever).

  </p>
</li>
<li><p>
  <strong><a name="print_tree_depths">print_tree_depths</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintTreeDepths">PrintTreeDepths</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
    Reconstruction of phylogenetic trees.

  </p>
</li>
<li><p>
  <strong><a name="print_variance_data">print_variance_data</a></strong>
  <i>[string fname="variance.dat"]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="PrintVarianceData">PrintVarianceData</a></strong>
  <i>[string fname="variance.dat"]</i>
  </p>
  <p>
    Print all of the variances of the average population statistics.

  </p>
</li>
<li><p>
  <strong><a name="print_viable_tasks_data">print_viable_tasks_data</a></strong>
  <i>[string fname="viable_tasks.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="PrintViableTasksData">PrintViableTasksData</a></strong>
  <i>[string fname="viable_tasks.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="SetVerbose">SetVerbose</a></strong>
  <i>[string verbosity=""]</i>
  </p>
  <p>
    Change the level of output verbosity.  Verbose messages will print all
  of the details of what is happening to the screen.  Minimal messages
  will only briefly state the process being run.  Verbose messages are
  recommended if you're in interactive analysis mode. When no arguments
  are supplied, action will toggle between NORMAL and ON.
  <br />Levels: SILENT, NORMAL, ON, DETAILS, DEBUG

  </p>
</li>
<li><p>
  <strong><a name="task_snapshot">task_snapshot</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="test_dom">test_dom</a></strong>
  <i>[string fname="dom-test.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="TestDominant">TestDominant</a></strong>
  <i>[string fname="dom-test.dat"]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="VERBOSE">VERBOSE</a></strong>
  <i>[string verbosity=""]</i>
  </p>
  <p>
  
  </p>
</li>
</UL>
<p>&nbsp;</p>
<h2><a name="SaveLoadActions">Save Load Actions</a></h2>

<p>
</p>
<UL>
<li><p>
  <strong><a name="detail_parasite_pop">detail_parasite_pop</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="detail_pop">detail_pop</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="detail_sex_pop">detail_sex_pop</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="dump_historic_pop">dump_historic_pop</a></strong>
  <i>[int back_dist=-1] [string fname=""]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="dump_historic_sex_pop">dump_historic_sex_pop</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="dump_pop">dump_pop</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="DumpPopulation">DumpPopulation</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="LoadClone">LoadClone</a></strong>
  <i>&lt;cString fname&gt;</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="load_clone">load_clone</a></strong>
  <i>&lt;cString fname&gt;</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="load_dump_file">load_dump_file</a></strong>
  <i>&lt;cString fname&gt; [int update=-1]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="LoadPopulation">LoadPopulation</a></strong>
  <i>&lt;cString fname&gt; [int update=-1]</i>
  </p>
  <p>
    Sets up a population based on a save file such as written out by
  SavePopulation. It is also possible to append a history file to the
  save file, in order to preserve the history of a previous run.

  </p>
</li>
<li><p>
  <strong><a name="save_clone">save_clone</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
    <font color="red">DEPRECATED</font>

  </p>
</li>
<li><p>
  <strong><a name="SaveClone">SaveClone</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
    Save a clone of this organism to the file specified; if no filename is
  given, use the name clone.<em>update</em>.  The update number allows regular
  clones with distinct filenames to be saved with the same periodic event.
  Running <kbd>avida -l filename</kbd> will start an Avida population with the
  saved clone.  Note that a clone only consists of the genomes in
  the population, and their current state is lost, so the run may not
  proceed identically as to if it had continued as it was going.

  </p>
</li>
<li><p>
  <strong><a name="SaveHistoricPopulation">SaveHistoricPopulation</a></strong>
  <i>[int back_dist=-1] [string fname=""]</i>
  </p>
  <p>
    This action is used to output all of the ancestors of the currently
  living population to the file specified, or <kbd>historic-<em>update</em>.pop</kbd>.

  </p>
</li>
<li><p>
  <strong><a name="SaveHistoricSexPopulation">SaveHistoricSexPopulation</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="SaveParasitePopulation">SaveParasitePopulation</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
<li><p>
  <strong><a name="SavePopulation">SavePopulation</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
    Save the genotypes and lots of statistics about the population to the
  file specified; if not filename is given, use the name
  <kbd>detail-<em>update</em>.pop</kbd>.  As with clones, the update number
  allows a single event to produce many detail files.  The details are used
  to collect crossection data about the population.

  </p>
</li>
<li><p>
  <strong><a name="SaveSexPopulation">SaveSexPopulation</a></strong>
  <i>[string fname=""]</i>
  </p>
  <p>
  
  </p>
</li>
</UL>
<p>&nbsp;</p>
<h2><a name="CreateAction">Creating an Action</a></h2>

<p>
The action source code is contained in the source/action directory.  
Each of the individual <a href="#ActionCategories">action categories</a> has its own source code files (e.g. Landcape Actions are located in the LandscapeActions files).
</p>
<p>Each action is derrived from the cAction class.  Briefly, to get an action to work you must create a child class that has a Process and GetDescription function defined as well as a constructor.  You must also register this new class with the action library.</p>
<p>So, with that quick review of what must be done, here is a step by step guide to creating an action:
<ol>
	<li>Identify which of the action categories your action should be assigned to.  There are six different action categories described <a href="#ActionCategories">above</a>.  Each category has a similar means of creating a new action, but do note that some action commands are generated via macros defined at the top of the files.  For instance, in the PrintActions file, you will notice a number of STATS_OUT_FILE macros being used to generate rather repetitively coded standard output files.&nbsp;</p></li>
	<li><p>Create a new class in the file that follows proper naming conventions.  Any class should begin with "cAction" and be followed by the name of the action command you will register with the library.  For instance, if we were to create a new command "MyAction", we'd name the class cActionMyAction.  Below is a stub for this new action class:</p>
			<p>
			<pre>
class cActionMyAction : public cAction
{
	private:
		// Private data members for this action
	public:
		cActionMyAction(cWorld* world, const cString& args) : cAction(world, args) { ; }
		
		static const cString GetDescription() { return "Arguments: My Arguments"; }
		
		void Process(cAvidaContext& ctx)
		{
			//Perform whatever processing is needed when the action is triggered.
		}
};
</pre></p>

	<li><p>Define the private data members, constructor, description string in GetDescription, and the Process function.  Any arguments that you specify after the action name in the events configuration will be passed to your new class via the args argument in the constructor.</p></li>
	<li><p>Register the new action with the action library.  At the bottom of each action definitions file, there are the commands that register the individual actions with the action library.  In the PrintActions.cc file, for instance, this function is called RegisterPrintActions.
	<p>  To register our example action "MyAction", we'd write:<p>
<p><pre>action_lib->Register&lt;cActionMyAction&gt;("MyAction");</pre><p>
	<li><p>Test your action.</p></li>
	<li><p>Add a ==== Action <i>action name</i> ==== line in
               source/utils/make_actions_html/actions_source_info file</p></li>
</ol>
</p>



<hr />
<p>
<a href="index.html">Return to the Index</a> &nbsp;|&nbsp;
<a href="events.html">The Events File</a> &nbsp;|&nbsp;
<a href="analyze.html">The Analyze File</a>
</p>

</body>
</html>
